core/platform/src/raw_syscalls.rs

```diff
@@ -0,0 +1,194 @@
+
+/// `RawSyscalls` allows a fake Tock kernel to be injected into components for
+/// unit testing. It is implemented by `libtock_runtime::TockSyscalls` and
+/// `libtock_unittest::FakeSyscalls`. Components should not use `RawSyscalls`
+/// directly; instead, use the `Syscalls` trait, which provides higher-level
+/// interfaces to the system calls.
+
+// RawSyscalls is designed to minimize the amount of handwritten assembly code
+// needed without generating unnecessary instructions. There are a few major
+// factors affecting its design:
+//     1. Most system calls only clobber r0-r4 (*), while yield has a far longer
+//        clobber list. As such, yield must have its own assembly
+//        implementation.
+//     2. The compiler is unable to optimize away unused arguments. For example,
+//        memop's "get process RAM start address" operation only needs r0 set,
+//        while memop's "break" operation needs both r0 and r1 set. If our
+//        inline assembly calls "get process RAM start address" but sets both r0
+//        and r1, the compiler doesn't know that r1 will be
+//        ignored so setting that register will not be optimized away. Therefore
+//        we want to set the minimum number of argument registers possible.
+//     3. The cost of specifying unused return registers is only that of
+//        unnecessarily marking a register as clobbered. Explanation: After
+//        inlining, an unused register is marked as "changed by the assembly"
+//        but can immediately be re-used by the compiler, which is the same as a
+//        clobbered register. System calls should generally be
+//        inlined -- and even if they aren't, the unused return values will
+//        probably be passed in caller-saved registers (this is true for the C
+//        ABI, so probably true for the Rust ABI), which are treated as
+//        clobbered regardless.
+//
+// (*) When this file refers to registers, it uses the same naming convention as
+// the Tock 2.0 syscalls TRD. Registers r0-r4 correspond to ARM registers r0-r4
+// and RISC-V registers a0-a4.
+//
+// Currently, yield takes exactly one argument, to specify what yield type to
+// do. Therefore we only need one raw yield call.
+//
+// Subscribe, command, read-write allow, and read-only allow all take four
+// argument types. Even when calling command IDs that have unused arguments, we
+// still need to clear the argument registers so as to avoid passing
+// confidential data to capsules (this is in line with Tock's threat model). As
+// such, four_arg_syscall() is used for all subscribe, command, read-only allow,
+// and read-write allow system calls.
+//
+// Memop takes 1 or 2 arguments (operation and an optional argument). Because it
+// is part of the core kernel, it is okay for us to leave arbitrary data in the
+// argument register for operations where the argument register is unused
+// (again, in line with Tock's threat model). Memop returns up to 2 return
```

Comment:
I guess by "returns [...] arguments" you are referring to the system call return value? As per the [current TRD state](https://github.com/tock/tock/blob/5e2e7db4d831661e9f4066cec18ac838d95882e8/doc/reference/trd-syscalls.md#46-memop-class-id-5), the "[...] success return type is Memop class system call specific", indicating that depending on the specifc Memop call, we might use the full 4 registers used for the various success return type variants.

I can imagine that this will not only be allowed theoretically, but that we might update Memop to take advantage of the gained registers and pass values commonly used together (such as the app's flash start and end address) in a single system call. We're going to touch & port Memop very soon, now that the required infrastructure is in place.

Reply:
I dug through all [existing memop types](https://github.com/tock/tock/blob/master/doc/syscalls/memop.md) and determined that none of them use any success types other than Success and Success with u32. My statement that "Memop returns up to 2 return arguments" is based on the existing memop types, and was not meant to be a theoretical limitation on all memop operations. I expect to redesign this interface (by adding r2 and/or r3 as return registers) if that changes.

I will update this comment to make that more clear.

Reply:
Great, thanks. I think it will be a good idea to document the return variants of the individual Memop operations once the design stabilizes, as part of the regular syscalls documentation in the kernel `docs/`. It's good to know that we are on the same page here and keeping this in mind.

Reply:
I just pushed commit [72a967](https://github.com/tock/libtock-rs/pull/261/commits/72a967363995cd0e805edfb35876aad9063a6d36), which updates the comments around memop. It should be more clear now that adding support for future memop operations may require changing clobbers.

I'm not too concerned about someone missing the clobber changes because adding a new return type would require changing the return type for `zero_arg_memop` and/or `one_arg_memop`. That return type change will force changes to the `RawSyscalls` implementation, and the obvious way to implement those changes is to implement correct output register specifications in the assembly.